File: src/test/java/io/prometheus/AdminEmptyPathTest.java
package io.prometheus;

import io.prometheus.client.CollectorRegistry;
import okhttp3.Request;
import okhttp3.Response;
import org.assertj.core.util.Lists;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeoutException;

import static io.prometheus.TestConstants.OK_HTTP_CLIENT;
import static java.lang.String.format;
import static java.util.concurrent.TimeUnit.SECONDS;
import static org.assertj.core.api.Assertions.assertThat;

public class AdminEmptyPathTest {

  private static final Logger logger = LoggerFactory.getLogger(AdminEmptyPathTest.class);

  private static Proxy PROXY = null;
  private static Agent AGENT = null;

  @BeforeClass
  public static void setUp()
      throws IOException, InterruptedException, TimeoutException {
    CollectorRegistry.defaultRegistry.clear();
    final List<String> args = Lists.newArrayList();
    args.add("-Dproxy.admin.port=8098");
    args.add("-Dproxy.admin.pingPath=\"\"");
    args.add("-Dproxy.admin.versionPath=\"\"");
    args.add("-Dproxy.admin.healthCheckPath=\"\"");
    args.add("-Dproxy.admin.threadDumpPath=\"\"");
    PROXY = TestUtils.startProxy(null, true, false, args);
    AGENT = TestUtils.startAgent(null, true, false, Collections.emptyList());

    AGENT.awaitInitialConnection(5, SECONDS);
  }

  @AfterClass
  public static void takeDown()
      throws InterruptedException, TimeoutException {
    PROXY.stopAsync();
    PROXY.awaitTerminated(5, SECONDS);
    AGENT.stopAsync();
    AGENT.awaitTerminated(5, SECONDS);
  }

  @Test
  public void proxyPingPathTest()
      throws Exception {
    assertThat(PROXY.getConfigVals().admin.port).isEqualTo(8098);
    assertThat(PROXY.getConfigVals().admin.pingPath).isEqualTo("");
    String url = format("http://localhost:%d/%s", PROXY.getConfigVals().admin.port, PROXY.getConfigVals().admin.pingPath);
    Request.Builder request = new Request.Builder().url(url);
    try (Response response = OK_HTTP_CLIENT.newCall(request.build()).execute()) {
      assertThat(response.code()).isEqualTo(404);
    }
  }

  @Test
  public void proxyVersionPathTest()
      throws Exception {
    assertThat(PROXY.getConfigVals().admin.port).isEqualTo(8098);
    assertThat(PROXY.getConfigVals().admin.versionPath).isEqualTo("");
    String url = format("http://localhost:%d/%s", PROXY.getConfigVals().admin.port, PROXY.getConfigVals().admin.versionPath);
    Request.Builder request = new Request.Builder().url(url);
    try (Response response = OK_HTTP_CLIENT.newCall(request.build()).execute()) {
      assertThat(response.code()).isEqualTo(404);
    }
  }

  @Test
  public void proxyHealthCheckPathTest()
      throws Exception {
    assertThat(PROXY.getConfigVals().admin.healthCheckPath).isEqualTo("");
    String url = format("http://localhost:%d/%s", PROXY.getConfigVals().admin.port, PROXY.getConfigVals().admin.healthCheckPath);
    Request.Builder request = new Request.Builder().url(url);
    try (Response response = OK_HTTP_CLIENT.newCall(request.build()).execute()) {
      assertThat(response.code()).isEqualTo(404);
    }
  }

  @Test
  public void proxyThreadDumpPathTest()
      throws Exception {
    assertThat(PROXY.getConfigVals().admin.threadDumpPath).isEqualTo("");
    String url = format("http://localhost:%d/%s", PROXY.getConfigVals().admin.port, PROXY.getConfigVals().admin.threadDumpPath);
    Request.Builder request = new Request.Builder().url(url);
    try (Response response = OK_HTTP_CLIENT.newCall(request.build()).execute()) {
      assertThat(response.code()).isEqualTo(404);
    }
  }
}

File: src/test/java/io/prometheus/AdminEmptyPathTest.kt
/*
 * Copyright © 2019 Paul Ambrose (pambrose@mac.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:Suppress("UndocumentedPublicClass", "UndocumentedPublicFunction")

package io.prometheus

import io.ktor.http.HttpStatusCode
import io.ktor.util.KtorExperimentalAPI
import io.prometheus.TestUtils.startAgent
import io.prometheus.TestUtils.startProxy
import io.prometheus.client.CollectorRegistry
import io.prometheus.common.simpleClassName
import io.prometheus.dsl.KtorDsl.blockingGet
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import mu.KLogging
import org.amshove.kluent.shouldEqual
import org.junit.AfterClass
import org.junit.BeforeClass
import org.junit.Test
import java.io.IOException
import java.util.concurrent.TimeoutException
import kotlin.time.ExperimentalTime
import kotlin.time.seconds

@KtorExperimentalAPI
@ExperimentalTime
@ExperimentalCoroutinesApi
class AdminEmptyPathTest {

    val proxyConfigVals = proxy.genericConfigVals.proxy

    @Test
    fun proxyPingPathTest() {
        proxyConfigVals.admin
            .also { admin ->
                admin.port shouldEqual 8098
                admin.pingPath shouldEqual ""

                blockingGet("${admin.port}/${admin.pingPath}".fixUrl()) { resp ->
                    resp.status shouldEqual HttpStatusCode.NotFound
                }
            }
    }

    @Test
    fun proxyVersionPathTest() {
        proxyConfigVals.admin
            .also { admin ->
                admin.port shouldEqual 8098
                admin.versionPath shouldEqual ""

                blockingGet("${admin.port}/${admin.versionPath}".fixUrl()) { resp ->
                    resp.status shouldEqual HttpStatusCode.NotFound
                }
            }
    }

    @Test
    fun proxyHealthCheckPathTest() {
        proxyConfigVals.admin
            .also { admin ->
                admin.healthCheckPath shouldEqual ""

                blockingGet("${admin.port}/${admin.healthCheckPath}".fixUrl()) { resp ->
                    resp.status shouldEqual HttpStatusCode.NotFound
                }
            }
    }

    @Test
    fun proxyThreadDumpPathTest() {
        proxyConfigVals.admin
            .also { admin ->
                admin.threadDumpPath shouldEqual ""

                blockingGet("${admin.port}/${admin.threadDumpPath}".fixUrl()) { resp ->
                    resp.status shouldEqual HttpStatusCode.NotFound
                }
            }
    }

    companion object : KLogging() {
        private lateinit var proxy: Proxy
        private lateinit var agent: Agent

        @JvmStatic
        @BeforeClass
        @Throws(IOException::class, InterruptedException::class, TimeoutException::class)
        fun setUp() {
            CollectorRegistry.defaultRegistry.clear()
            val args = listOf("-Dproxy.admin.port=8098",
                              "-Dproxy.admin.pingPath=\"\"",
                              "-Dproxy.admin.versionPath=\"\"",
                              "-Dproxy.admin.healthCheckPath=\"\"",
                              "-Dproxy.admin.threadDumpPath=\"\"")

            logger.info { "Starting ${proxy.simpleClassName} and ${agent.simpleClassName}" }
            runBlocking {
                launch(Dispatchers.Default) { proxy = startProxy(adminEnabled = true, argv = args) }
                launch(Dispatchers.Default) {
                    agent = startAgent(adminEnabled = true).apply { awaitInitialConnection(5.seconds) }
                }
            }
            logger.info { "Finished starting ${proxy.simpleClassName} and ${agent.simpleClassName}" }
        }

        @JvmStatic
        @AfterClass
        @Throws(InterruptedException::class, TimeoutException::class)
        fun takeDown() {
            runBlocking {
                for (service in listOf(proxy, agent)) {
                    logger.info { "Stopping ${service.simpleClassName}" }
                    launch(Dispatchers.Default) { service.stopSync() }
                }
            }
            logger.info { "Finished stopping ${proxy.simpleClassName} and ${agent.simpleClassName}" }
        }
    }
}
